src/lib.rs

```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
-
-
-        // ==== Constraints ====
+    trait Store for Module<T: Trait> as Forum {
+
+        /// Map category identifier to corresponding category.
+        pub CategoryById get(category_by_id) config(): map CategoryId => Category<T::BlockNumber, T::Moment, T::AccountId>;
+
+        /// Category identifier value to be used for the next Category created.
+        pub NextCategoryId get(next_category_id) config(): CategoryId;
+
+        /// Map thread identifier to corresponding thread.
+        pub ThreadById get(thread_by_id) config(): map ThreadId => Thread<T::BlockNumber, T::Moment, T::AccountId>;
+
+        /// Thread identifier value to be used for next Thread in threadById.
+        pub NextThreadId get(next_thread_id) config(): ThreadId;
+
+        /// Map post identifier to corresponding post.
+        pub PostById get(post_by_id) config(): map PostId => Post<T::BlockNumber, T::Moment, T::AccountId>;
+
+        /// Post identifier value to be used for for next post created.
+        pub NextPostId get(next_post_id) config(): PostId;
+
+        /// Account of forum sudo.
+        pub ForumSudo get(forum_sudo) config(): Option<T::AccountId>;
 
+        // === Add constraints here ===
+
+        // Will add all the constrainst here later!
+
+    }
+    /*
+    JUST GIVING UP ON ALL THIS FOR NOW BECAUSE ITS TAKING TOO LONG
+    Review : https://github.com/paritytech/polkadot/blob/620b8610431e7b5fdd71ce3e94c3ee0177406dcc/runtime/src/parachains.rs#L123-L141
+
+    add_extra_genesis {
+
+        // Explain why we need to put this here.
+        config(initial_forum_sudo) : Option<T::AccountId>;
+
+        build(|
+            storage: &mut generator::StorageOverlay, 
+            _: &mut generator::ChildrenStorageOverlay,
+            config: &GenesisConfig<T>
+            | {
+
+
+			if let Some(account_id) = &config.initial_forum_sudo {
+                println!("{}: <ForumSudo<T>>::put(account_id)", account_id); 
+                <ForumSudo<T> as generator::StorageValue<_>>::put(&account_id, storage);
+            }
+		})
     }
+    */
 }
 
 decl_event!(
-    /// Events are a simple means of reporting specific conditions and
-    /// circumstances that have happened that users, Dapps and/or chain explorers would find
-    /// interesting and otherwise difficult to detect.
     pub enum Event<T>
     where
-        B = BalanceOf<T>,
+        <T as system::Trait>::AccountId,
     {
-        // Just a normal `enum`, here's a dummy event to ensure it compiles.
-        /// Dummy event, just here so there's a generic type that's used.
-        Dummy(B),
+        /// A category was introduced
+        CategoryCreated(CategoryId),
+
+        /// A category with given id was updated. 
+        /// The second argument reflects the new archival status of the category, if changed.
+        /// The third argument reflects the new deletion status of the category, if changed.
+        CategoryUpdated(CategoryId, Option<bool>, Option<bool>),
```

Comment:
Did you consider to create more granular events like?:
- `CategoryArchived`
- `CategoryUnarchived`
- `CategoryDeleted`
- `CategoryUndeleted`

Having a few sequential parameters of the same type (Option<bool>) could be confusing when consuming the event because event parameters are not named.

Reply:
I think that is better solved by just having proper documentation of events, which we currently don't have. One could make a similar argument to adding four different dispatchables also, but I don't think its worth going back on now.